src/core/server/http/http_server.ts

```diff
@@ -163,43 +163,54 @@ export class HttpServer {
         const validate = isSafeMethod(route.method) ? undefined : { payload: true };
         const { authRequired, tags, body = {}, timeout } = route.options;
         const { accepts: allow, maxBytes, output, parse } = body;
-        // Hapi does not allow timeouts on payloads to be specified for 'head' or 'get' requests
-        const payloadTimeout = isSafeMethod(route.method) || timeout == null ? undefined : timeout;
 
         const kibanaRouteState: KibanaRouteState = {
           xsrfRequired: route.options.xsrfRequired ?? !isSafeMethod(route.method),
         };
 
+        // Works around https://github.com/hapijs/hapi/issues/4122 until v20
+        const fakeSocketTimeout = timeout?.payload ? timeout.payload + 1 : undefined;
+
         this.server.route({
           handler: route.handler,
           method: route.method,
           path: route.path,
           options: {
             auth: this.getAuthOption(authRequired),
             app: kibanaRouteState,
+            ext: {
+              onPreAuth: {
+                method: (request, reply) => {
+                  if (timeout?.idleSocket) {
+                    request.raw.req.socket.setTimeout(timeout?.idleSocket);
+                  } else if (fakeSocketTimeout) {
+                    request.raw.req.socket.setTimeout(this.config!.socketTimeout || 0);
+                  }
+
+                  return reply.continue;
```

Comment:
nit: in Hapi v17+, the second argument here is the response toolkit and is usually named `h` instead of `reply`. Example: https://hapi.dev/api?v=17.9.3#-serverextevent-method-options

```suggestion
                method: (request, h) => {
                  if (timeout?.idleSocket) {
                    request.raw.req.socket.setTimeout(timeout?.idleSocket);
                  } else if (fakeSocketTimeout) {
                    request.raw.req.socket.setTimeout(this.config!.socketTimeout || 0);
                  }

                  return h.continue;
```